src/libsyntax/ast.rs

```diff
@@ -1746,13 +1730,21 @@ impl StructFieldKind {
     }
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
+pub enum VariantKind {
+    Dict,
```

Comment:
What is `Dict`? Short for dictionary? we don't generally use that terminology in Rust. This enum is for struct kinds now, right? Not variant kinds. Apart from naming I find this enum a little bit weird - what kind of struct it is really just depends on the kind of the fields, which is encoded elsewhere, so there is an implicit constraint between the field kinds and the struct kinds. In fact, there is no reason why a struct should not mix kinds of fields - especially in the HIR.

Perhaps a better way to organise things is to have a DataKind enum with variants for tuple, and struct, including the Vec of fields (un-named and named fields, respectively) - unit = tuple with no fields, unit with braces = struct with no fields. StructDef could be renamed NamedDataDef since it now covers way more than just structs.